server/src-lib/Hasura/Server/Init.hs

```diff
@@ -1061,3 +1069,52 @@ serveOptionsParser =
   <*> parseEnabledLogs
   <*> parseLogLevel
   <*> parsePlanCacheSize
+
+downgradeOptionsParser :: Parser DowngradeOptions
+downgradeOptionsParser = 
+    DowngradeOptions
+    <$> option (maybeReader readVersion)
+        ( long "to" <>
+          metavar "<VERSION>" <>
+          help "The target schema version (e.g. v1.0.0-beta.1)"
+        )
+    <*> switch
+        ( long "dryRun" <>
+          help "Don't run any migrations, just print out the SQL."
+        )
+  where
+    -- This implements the mapping between application versions
+    -- and catalog schema versions.
+    readVersion "v1.0.0-alpha0"  = Just "1"
```

Comment:
I actually think having the mapping here is strictly worse than keeping it in documentation, because it's easier to forget to update it, and the user won't have any way to downgrade in that case. At least using catalog versions, we can tell them the exact command to use. I'm happy to hear alternatives if anyone has ideas.

Reply:
Why do you think it’s easier to forget to update this mapping than one in the documentation? I don’t see why one would be more obvious than the other.

In any case, I think being able to treat catalog versions as an implementation detail our users don’t have to worry about is a big enough win that I lean toward this approach, anyway. However, it does make me want two other features:

  1. I’d like to be able to print the list of versions `graphql-engine` knows about somehow. Otherwise, if I enter something like `1.1.0-beta.1` (leaving off the `v`) I just get a cryptic “cannot parse value” error, which is very unhelpful.

     I think the error message for supplying an invalid version should be improved to either list the known versions or tell people how to list them, and there should be something in the `--help` output that says how to list them. (We probably don’t want to list them all in the `--help` text all the time, since it would be a lot of output. But maybe we could show an example instead, and then we could probably get away with only printing out all the known versions if someone supplies an invalid version.)

  2. I’d like to be able to specify a catalog version by specifying a `--catalog-version` flag or something like that. This is useful both for developers working on the server and users trying out PR builds. In the former case, we know the catalog versions, and in the latter, we’re already interacting with users directly, so we can just tell them what to run. Since this would be so easy to add, it seems worthwhile!

Reply:
@lexi-lambda How's this?